### PR TITLE
Add per-event overloads to ``EventManager.connect()``

### PR DESCRIPTION
**Pull request by adamtheturtle**
This brings the changes in https://github.com/sphinx-doc/sphinx/pull/12784 to this function.

### Feature or Bugfix
- Feature

### CODE REVIEW

**Comment by AA-Turner** [comment]
Is there a way we can avoid the duplication? Is ParamSpec applicable here?

A

**Comment by adamtheturtle** [comment]
> Is there a way we can avoid the duplication? Is ParamSpec applicable here?

I'm not aware of a way. I think `ParamSpec` is not applicable, but I have not ever used it. In particular, I think that any attempt at de-duplication might be quite hacky not least because the `@overload`s are not directly copy-pasted:

* They have different parameter names (`event` versus `name`)
* `priority` has a default in `sphinx.application.Sphinx.connect`, but not in `sphinx.events.EventManager.connect`

File: sphinx/events.py [hunk 1]
```diff
@@ -7,17 +7,26 @@
 
 from collections import defaultdict
 from operator import attrgetter
-from typing import TYPE_CHECKING, Any, NamedTuple
+from typing import TYPE_CHECKING, Any, Literal, NamedTuple, overload
```
Comment (AA-Turner) [inline sphinx/events.py]:
```suggestion
from typing import TYPE_CHECKING, overload
```

File: sphinx/events.py [hunk 2]
```diff
@@ -7,17 +7,26 @@
...
 
 from sphinx.errors import ExtensionError, SphinxError
 from sphinx.locale import __
 from sphinx.util import logging
 from sphinx.util.inspect import safe_getattr
 
 if TYPE_CHECKING:
-    from collections.abc import Callable
+    from collections.abc import Callable, Iterable, Sequence, Set
+    from pathlib import Path
```
Comment (AA-Turner) [inline sphinx/events.py:19]:
```suggestion
    from pathlib import Path
    from typing import Any, Literal, NamedTuple
```

**Pull request merged**